control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -96,38 +96,74 @@ func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	}
 	r.Log.Info("retrieved Service", "name", req.Name, "ns", req.Namespace)
 
-	workloadSelector, err := r.getWorkloadSelectorFromEndpoints(ctx, &ClientPodFetcher{client: r.Client}, &endpoints)
+	consulSvc, err := r.getConsulService(ctx, &ClientPodFetcher{client: r.Client}, service, endpoints)
 	if err != nil {
-		errs = multierror.Append(errs, err)
+		r.Log.Error(err, "failed to build Consul service resource", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
 	}
 
-	// If we have at least one mesh-injected pod targeted by the service, register it in Consul.
+	// If we don't have at least one mesh-injected pod targeted by the service, do not register the service.
 	//TODO: Register service with mesh port added if global flag for inject is true,
 	// even if Endpoints are empty or have no mesh pod, iff. the service has a selector.
 	// This should ensure that we don't target kube or consul (system) services.
-	if workloadSelector != nil {
-		//TODO: Maybe check service-enable label here on service/deployments/other pod owners
-		if err = r.registerService(ctx, resourceClient, service, workloadSelector); err != nil {
-			// We could be racing with the namespace controller.
-			// Requeue (which includes backoff) to try again.
-			if common.ConsulNamespaceIsNotFound(err) {
-				r.Log.Info("Consul namespace not found; re-queueing request",
-					"service", service.GetName(), "ns", req.Namespace,
-					"consul-ns", r.getConsulNamespace(req.Namespace), "err", err.Error())
-				return ctrl.Result{Requeue: true}, nil
-			}
-			errs = multierror.Append(errs, err)
+	if consulSvc.Workloads == nil {
+		return ctrl.Result{}, nil
+	}
```

Comment:
This replaces the check above - where we used to create just the workload selector here and let the `registerService` call create the service for us, we now do the heavy lifting inside `getConsulService `.

The surrounding code in this `Reconcile` func is effectively unchanged.